### PR TITLE
Clean up DynamicDependencyAttribute usages

### LINKED ISSUE #37837: Clean up DynamicDependencyAttribute usages
- Some of these attributes are no longer needed because the linker's pattern detection will be able to understand the code and keep the appropriate members. This [change](https://github.com/sbomer/runtime/compare/dynamicDependency...sbomer:removeDynamicDependency?expand=1) shows the ones that should be removable.
- Some of them should be replaced with `DynamicallyAccessedMembersAttribute`. See https://github.com/dotnet/runtime/pull/37780/commits/e11faf6e19162d5ebf91f93d9fbb9dafc0f317af.

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/37837.

The first commit adds support for having common source files that can be shared between multiple trimming tests. I needed this for the CookieExtensions tests. Thanks @joperezr .

### CODE REVIEW

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs [hunk 1]
```diff
@@ -966,6 +969,8 @@ public static PropertyFetch FetcherForProperty(Type? type, string propertyName)
 
                     private sealed class RefTypedFetchProperty<TObject, TProperty> : PropertyFetch
                     {
+                        [DynamicDependency("get_TraceStateString", typeof(Activity))]
```
Comment (eerhardt) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:972]: I'm not understanding the reasoning behind these DynamicDependencies. Can you explain?

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs [hunk 2]
```diff
@@ -949,6 +946,12 @@ public static PropertyFetch FetcherForProperty(Type? type, string propertyName)
                                 Logger.Message($"Property {propertyName} not found on {type}");
                                 return new PropertyFetch(type);
                             }
+                            // Delegate creation below is incompatible with static properties.
```
Comment (eerhardt) [inline src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:949]: Is this an existing bug in the code? This doesn't seem related to the linker.

File: src/libraries/System.Net.HttpListener/tests/TrimmingTests/CookieToServerStringTest.cs [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace CookieToServerStringTest
+{
+    /// <summary>
+    /// Tests that the System.Net.CookieExtensions.ToServerString()
```
Comment (eerhardt) [inline src/libraries/System.Net.HttpListener/tests/TrimmingTests/CookieToServerStringTest.cs:17]: What about the other reflection usages?

File: src/libraries/System.Net.HttpListener/tests/TrimmingTests/CookieToServerStringTest.cs [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    /// method works as expected when used in a trimmed application.
+    /// </summary>
+    internal class Program
```
Comment (eerhardt) [inline src/libraries/System.Net.HttpListener/tests/TrimmingTests/CookieToServerStringTest.cs:20]: Is this a minimal program to test this scenario? It seems like a lot of code, can any of it be removed?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactory.cs [hunk 5]
```diff
@@ -25,23 +25,6 @@ public override bool CanConvert(Type typeToConvert)
             return typeof(IEnumerable).IsAssignableFrom(typeToConvert);
         }
 
-        [DynamicDependency("#ctor", typeof(ArrayConverter<,>))]
```
Comment (eerhardt) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactory.cs:28]: Do we have tests for these?

File: src/libraries/System.Net.HttpListener/tests/TrimmingTests/CookieToServerStringTest.cs (hunk 4 above)
Comment (layomia) [reply src/libraries/System.Net.HttpListener/tests/TrimmingTests/CookieToServerStringTest.cs]: Yup, I adapted much of this from existing tests. I've trimmed out unneeded parts.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactory.cs (hunk 5 above)
Comment (layomia) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactory.cs]: As discussed offline, I'll make the JSON changes in a dedicated PR.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:972]: The `FetcherForProperty` method reflects over properties on `System.Diagnostics.Activity`, depending on what properties are passed to an `EventListener`, e.g. see [this usage](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/tests/TrimmingTests/EventSourcePropertyFetchTest.cs#L23-L47). I believe the getters for these properties are used to report activity on whatever the listener is listening for. 

The [example](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/tests/TrimmingTests/EventSourcePropertyFetchTest.cs#L23-L47) passes in all non-static* properties on `typeof(Activity)` to a listener, and it turns out that the getters for `TraceStateString` and `IsAllDataRequested` are the only two that are trimmed out. This would cause a [NullReferenceException ](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L977)in a trimmed app with this usage pattern.

---
*A problem with static properties is discussed here - https://github.com/dotnet/runtime/pull/38216#discussion_r443764227.

Comment (MichalStrehovsky) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:972]: This looks like an annotation that makes the trimming test pass, but doesn't actually make DiagnosticSource trimming safe.

There seems to be code in DiagnosticSource that reflects on properties passed as a text string into the source. The text string is `activityProps` in the `EventSourcePropertyFetchTest.cs`.

If we were to add a property on a class there that is not otherwise used, the reflection is going to crash and fail.

I think if we were to replace the 

```csharp
var obj = new { DummyProp = "val" };
```

in the test with

```csharp
var obj = new NeverUsedProperties();
// class NeverUsedProperties { public string DummyProp { get; set; } }
```

This is actually going to crash because linker will trim `DummyProp` away, but the `activityProps` string is still going to refer to it.

We might want to just annotate DiagnosticSourceEventSource as not trimming safe and revisit in .NET 6.

Comment (MichalStrehovsky) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:972]: Well, maybe it won't crash, just silently not work, because the activityProps string is already referencing a bunch of non-existent properties that DiagnosticSource is complaining about:

```
Property Links not found on System.Diagnostics.Activity
Property IsAllDataRequested not found on System.Diagnostics.Activity
Property Kind not found on System.Diagnostics.Activity
Property DisplayName not found on System.Diagnostics.Activity
Property Source not found on System.Diagnostics.Activity
Property Events not found on System.Diagnostics.Activity
Property Context not found on System.Diagnostics.Activity
```

Comment (layomia) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:972]: FWIW - In an earlier iteration of this PR, I placed `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)]` on the `type` parameter in [this method](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L895), which I thought would preserve the needed properties and their accessors. That didn't work. Is the expected behavior with `DynamicallyAccessedMemberTypes.PublicProperties` that the accessors will be preserved?

File: eng/testing/linker/trimmingTests.targets [hunk 6]
```diff
@@ -7,7 +7,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <TestConsoleAppSourceFiles Include="$(MSBuildProjectDirectory)\*.cs" />
+    <TestConsoleAppSourceFiles Condition="'@(TestConsoleAppSourceFiles)' == ''" Include="$(MSBuildProjectDirectory)\*.cs" />
```
Comment (joperezr) [inline eng/testing/linker/trimmingTests.targets:10]: not sure why this shows as a diff since this is the state this line is in master today...

File: src/libraries/System.Data.Common/src/System/Data/SQLTypes/SqlXml.cs [hunk 7]
```diff
@@ -128,7 +128,6 @@ internal static XmlReader CreateSqlXmlReader(Stream stream, bool closeInput = fa
 
         private static MethodInfo CreateSqlReaderMethodInfo
         {
-            [DynamicDependency("CreateSqlReader", typeof(System.Xml.XmlReader))]
```
Comment (joperezr) [inline src/libraries/System.Data.Common/src/System/Data/SQLTypes/SqlXml.cs:131]: Is this change because the linker can now understand this use? I thought that reflecting on a method on a different assembly would still not always be detected.

File: src/libraries/System.Linq.Expressions/tests/TrimmingTests/System.Linq.Expressions.TrimmingTests.proj [hunk 8]
```diff
@@ -0,0 +1,6 @@
+
+<Project DefaultTargets="Build">
```
Comment (joperezr) [inline src/libraries/System.Linq.Expressions/tests/TrimmingTests/System.Linq.Expressions.TrimmingTests.proj:2]: NIT: Empty line above this.

File: src/libraries/System.Net.HttpListener/tests/TrimmingTests/System.Net.HttpListener.TrimmingTests.proj [hunk 9]
```diff
@@ -0,0 +1,18 @@
+
+<Project DefaultTargets="Build">
```
Comment (joperezr) [inline src/libraries/System.Net.HttpListener/tests/TrimmingTests/System.Net.HttpListener.TrimmingTests.proj:2]: NIT: extra empty line above this.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:972]: > Is the expected behavior with DynamicallyAccessedMemberTypes.PublicProperties that the accessors will be preserved?

You need to add annotations all the way up until you get to a concrete type. The code that calls into `PropertyFetch.FetcherForProperty` doesn't call it with a concrete type:

https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L858-L864

Here the linker has no way of understanding what `obj?.GetType()` will return since `obj` is of type `object`.

I agree it makes sense to mark `DiagnosticSourceEventSource` as not trimming safe for .NET 5 and log an issue to address it in the future.

Comment (MichalStrehovsky) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:972]: Annotating `type` would do the right thing for the code in that method, but I assume you got a warning from [this line](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L862) because linker wouldn't be able to figure out where that `objType` comes from.

I tried to manually trace that back, but I gave up after I saw the source of the `objType` ([this line](https://github.com/dotnet/runtime/blob/923c57b38b1da3f1bf8e3108c10cca3cfd5974e0/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs#L859)) requires tracing through a KeyValuePair, delegate, interface call, and I still couldn't figure out what types could be stored in the `object`-typed parameter. The only safe thing to do for linker in that spot would be to preserve all property on everything deriving from `System.Object` and that's kind of counter productive to trimming.

File: src/libraries/System.Data.Common/src/System/Data/SQLTypes/SqlXml.cs (hunk 7 above)
Comment (MichalStrehovsky) [reply src/libraries/System.Data.Common/src/System/Data/SQLTypes/SqlXml.cs:131]: The bug you're referring to is this one: https://github.com/mono/linker/issues/943

It affects situation when linker analyzes a reflection pattern, finds out type X from assembly Y is needed, but assembly Y wasn't in the assembly closure linker initially started looking at (i.e. we would need to pull in a new assembly).

It can only happen in `Type.GetType("SomeType, SomeAssembly")` - the pattern in this file is safe from that bug because there's a typeof in the method body.

File: eng/testing/linker/trimmingTests.targets (hunk 6 above)
Comment (layomia) [reply eng/testing/linker/trimmingTests.targets:10]: Hmm I wasn't synced to master when I manually made this change. Hopefully, the diff won't cause any adverse effects.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs (hunk 1 above)
Comment (layomia) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:972]: I see.

I'll pull out related changes from this PR. I've logged an issue to mark the method as unsafe for trimming - https://github.com/dotnet/runtime/issues/38404. Will address in a follow up.

File: src/libraries/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs [hunk 10]
```diff
@@ -42,7 +40,6 @@ private enum CookieVariant
 
         private static Func<Cookie, CookieVariant> s_getVariantFunc;
 
-        [DynamicDependency("get_Variant", typeof(Cookie))]
```
Comment (layomia) [inline src/libraries/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs:45]: I pinged @scalablecory offline for a suggestion for how to test this change. I'll follow up with a PR when we figure one out.

File: src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs (hunk 2 above)
Comment (layomia) [reply src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/DiagnosticSourceEventSource.cs:949]: I logged https://github.com/dotnet/runtime/issues/38406.

File: eng/testing/linker/trimmingTests.targets (hunk 6 above)
Comment (layomia) [reply eng/testing/linker/trimmingTests.targets]: Fixed after moving my changes on top of master.

**Review by eerhardt — APPROVED** [review]
LGTM

File: src/libraries/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs (hunk 10 above)
Comment (layomia) [reply src/libraries/System.Net.HttpListener/src/System/Net/Windows/CookieExtensions.cs:45]: Logged issue - https://github.com/dotnet/runtime/issues/38447.
